### PR TITLE
Add search keywords to the class reference

### LINKED ISSUE #5085: Add keywords for easier searching of methods/properties/… in the class reference
### Describe the project you are working on

The Godot editor :slightly_smiling_face: 

### Describe the problem or limitation you are having in your project

Finding properties/methods in the class reference can be difficult if you don't know the exact name of the property/method. See https://github.com/godotengine/godot-proposals/issues/5077 where this issue was encountered.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

Add keywords for easier searching of methods/properties/… in the class reference.

This is something I originally proposed in https://github.com/godotengine/godot/issues/29286#issuecomment-497135808 and https://github.com/godotengine/godot/issues/29286, but it seems I forgot to open a proposal about it until now.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

- Add an *optional* `keywords="keyword1, keyword2, ..."` attribute on `property`/`method`/`signal`/`constant`/`theme_item`/`enum`/`annotation` tags. This attribute won't be generated by `--doctool`, so it'll be up for contributors to add it where needed. (For consistency, tt should always be added as the last attribute.)
  - The `keywords` tag is probably not relevant for constructors and operators.
- Modify `--doctool` to preserve the `keywords` attribute on relevant tags.
- Update the [XML schema](https://github.com/godotengine/godot/blob/master/doc/class.xsd) to allow the optional presence of a `keywords` attribute on relevant tag.

The online class reference may be able to benefit from this by collecting all keyword aliases for each class, then throwing them all under a `<meta keywords="...">` tag using the following reStructuredText at the top:

```rst
.. meta::
    :keywords: keyword1, keyword2, ...
```

This `keywords` meta tag is ignored by search engines like Google, but the internal search bar will make use of them.

### If this enhancement will not be used often, can it be worked around with a few lines of script?

No.

### Is there a reason why this should be core and not an add-on in the asset library?

This is about improving documentation usability.

### PR DESCRIPTION
**Pull request by RedMser**
Fixes https://github.com/godotengine/godot-proposals/issues/5085

Allows for finding methods, properties, signals, constants, theme items and annotations more easily.

- Allow "keywords" attribute in aforementioned locations in the class reference XMLs.
- Extends doctool, to preserve these attributes.
- Update the XSD schema for the class reference.
- Update the RST generator to include a meta tag for class keywords.
- Update editor help search to allow filtering for these keywords.
- Add some exemplary keywords attributes in the class reference, but more can be added in future PRs. They are merely included to not be forgotten in the proposal, and to be able to test the feature.

image

# Ideas and Future

- Should it be possible to add a `@keywords ...` comment in GDScript to define keywords for script classes and members? Implementation would look like #78941
- Idea for the future: add keyword search to GDScript autocompletion
- Idea for the future: add keyword search to the "create new node/resource" dialog
- Render the keywords list in a few places, such as
  - Next to a matching search result (so you know why this result is showing)
  - In the documentation (both RST and in the editor), to get a full list of related keywords

### CODE REVIEW

File: editor/editor_help_search.cpp [hunk 1]
```diff
@@ -40,6 +40,7 @@ void EditorHelpSearch::_update_icons() {
 	search_box->set_right_icon(results_tree->get_theme_icon(SNAME("Search"), SNAME("EditorIcons")));
 	search_box->set_clear_button_enabled(true);
 	search_box->add_theme_icon_override("right_icon", results_tree->get_theme_icon(SNAME("Search"), SNAME("EditorIcons")));
+	keywords_button->set_icon(results_tree->get_theme_icon(SNAME("AnimationTrackList"), SNAME("EditorIcons")));
```
Comment (RedMser) [inline editor/editor_help_search.cpp:43]: Might need a better choice of icon, since it is visually very similar to the "Show Hierarchy" icon.

**Comment by YuriSizov** [comment]
> Update editor help search with a toggle button to allow filtering for these keywords.

What's the use case for not wanting to search by keywords?

**Comment by RedMser** [comment]
> What's the use case for not wanting to search by keywords?

Right now, the keywords do not show up anywhere in the UI, so I'm not sure if power users are gonna be happy that seemingly unrelated results suddenly show up in the search results.

If I had more time to work on this PR, I'd show the keyword that matched next to the search result, with greyed out text. Might take me a bit before I can get that implemented.

But I'm fine with taking the button out if there's no apparent need for it.

**Comment by YuriSizov** [comment]
> I'm not sure if power users are gonna be happy that seemingly unrelated results suddenly show up in the search results

Well the point of having keywords would be to make sure we show more related results than we do now. Hiding them behind a toggle makes them harder to access and less useful.

**Comment by RedMser** [comment]
Removed the toggle so keyword searching is always enabled now.

**Review by Calinou — APPROVED** [review]
Tested locally, it works. Code looks good to me.

There's a small usability issue however:

- When you search for a node directly, it's automatically focused:

image

- When you search for a keyword, the node that matches the keyword isn't automatically selected:

image

Getting this working is important to reduce the number of keystrokes you need to open the documentation page, so I suggest looking into this if possible.

**Comment by RedMser** [comment]
@Calinou Done! Updated `_match_item` and calling it for each keyword.

**Comment by YuriSizov** [comment]
Reducing the weight makes sense, as it's an indirect match.

**Comment by akien-mga** [comment]
I'm not too fond of having keywords as XML tag attributes.

Attributes are usually things generated from the codebase, and human-written content should be filled within tags (e.g. `<description>Human written</description>`).

So it might make more sense IMO to add an optional `<keywords>` tag, like we do with `<link>` for the `<tutorials>` section.

**Comment by RedMser** [comment]
> So it might make more sense IMO to add an optional `<keywords>` tag

@akien-mga While I agree with your reasoning, it seems like it would have to be a breaking XSD schema change.
`<member>`, `<constant>` and `<theme_item>` are tags that [only have a string content](https://github.com/godotengine/godot/blob/master/doc/classes/PhysicalBone2D.xml#L28-L30), so all of their descriptions would have to be moved into a `<description>` element.

Maintaining a backwards compatible parser sounds problematic (especially now that the version attribute is gone) and it would make new documentation features hard to discover.
Going through with the breaking change would (besides a large diff) break doctools of user-maintained modules, which might not be too bad?

Let me know what you think!

**Comment by akien-mga** [comment]
I don't understand, how would `<keywords>3d, spatial</keywords>` differ from the existing optional `<link>https://example.com</link>` tag?

**Review by Calinou — APPROVED** [review]
Tested locally (rebased on top of `master` 57919beb05687cf2ee80ed0d3ccac6445c7edea9), it works as expected with correct focus behavior now.

I can do a pass over the class reference to add keywords once this PR is merged.

**Comment by Mickeon** [comment]
Would be nice to address the conflicts. Reception on this seems fairly positive. Although the current implementation seems out of reach for translators, if that was ever a concern.

**Comment by RedMser** [comment]
Rebased onto latest master. Thanks for letting me know! ^^

> Although the current implementation seems out of reach for translators, if that was ever a concern.

Since the API is not translated (so method names, property names, etc.), I don't think it matters. You wouldn't search for a Spanish keyword when all the possible results are only ever in English.

**Review by Mickeon — APPROVED** [review]
Took a deep look. You can get a lot of mileage out of this feature in the future.

It is kind of confusing to get search results from terms that are not directly visible, although that can be figured out later.

**Comment by Calinou** [comment]
Note for the future: Vector2/3's `limit_length` could have `truncate` keyword as mentioned in https://github.com/godotengine/godot-proposals/discussions/9020.

File: editor/editor_help_search.cpp [hunk 2]
```diff
@@ -711,43 +739,46 @@ TreeItem *EditorHelpSearch::Runner::_create_class_item(TreeItem *p_parent, const
 	}
 
 	_match_item(item, p_doc->name);
+	for (String keyword : p_doc->keywords.split(",")) {
```
Comment (AThousandShips) [inline editor/editor_help_search.cpp:742]:
```suggestion
	for (const String &keyword : p_doc->keywords.split(",")) {
```

File: editor/editor_help_search.cpp [hunk 3]
```diff
@@ -600,7 +610,20 @@ bool EditorHelpSearch::Runner::_match_string(const String &p_term, const String
 	}
 }
 
-void EditorHelpSearch::Runner::_match_item(TreeItem *p_item, const String &p_text) {
+bool EditorHelpSearch::Runner::_match_keywords(const String &p_term, const String &p_keywords) const {
+	for (String keyword : p_keywords.split(",")) {
```
Comment (AThousandShips) [inline editor/editor_help_search.cpp:614]:
```suggestion
	for (const String &keyword : p_keywords.split(",")) {
```

File: editor/editor_help_search.cpp [hunk 4]
```diff
@@ -583,15 +584,24 @@ void EditorHelpSearch::Runner::_match_method_name_and_push_back(Vector<DocData::
...
-		if (!_match_string(terms[i], name)) {
+		if (!_match_string(terms[i], p_name)) {
 			return false;
 		}
 	}
 	return true;
 }
 
+bool EditorHelpSearch::Runner::_all_terms_in_keywords(const String &p_keywords) const {
+	for (String keyword : p_keywords.split(",")) {
```
Comment (AThousandShips) [inline editor/editor_help_search.cpp:597]:
```suggestion
	for (const String &keyword : p_keywords.split(",")) {
```

File: editor/editor_help_search.cpp [hunk 5]
```diff
@@ -774,6 +805,9 @@ TreeItem *EditorHelpSearch::Runner::_create_member_item(TreeItem *p_parent, cons
 	}
 
 	_match_item(item, p_name);
+	for (String keyword : p_keywords.split(",")) {
```
Comment (AThousandShips) [inline editor/editor_help_search.cpp:808]:
```suggestion
	for (const String &keyword : p_keywords.split(",")) {
```

**Review by AThousandShips — APPROVED** [review]
Looks good otherwise

**Comment by akien-mga** [comment]
> It is kind of confusing to get search results from terms that are not directly visible, although that can be figured out later.

Yeah that's my main concern with the current implementation. I'd like to see something that makes it obvious to the user why the filtered entry match their search. There could e.g. be a "*(keyword: ragdoll)*" added after the node name in the list, in the OP's example.

It shouldn't necessarily be blocking for this PR though, it could be worked on in a follow-up, possibly by someone else if wanted.

**Comment by akien-mga** [comment]
Thanks!

**Pull request merged**